weldx/measurement.py

```diff
@@ -47,55 +55,459 @@ class DataTransformation:
+        """Raise a `KeyError` if the specified node does not exist.
 
         Parameters
         ----------
         node :
-            Name of the new node
-        parent_node :
-            Name of the parent node
-        node_label :
-            Displayed name of the node
-        position :
-            Position of the node
-        container :
-            Tuple of containers that should be updated.
+            Name of the node that should be searched for
 
         """
-        [graph, node_list, labels, positions] = container
-        graph.add_node(node)
-        node_list.append(node)
-        labels[node] = node_label
-        positions[node] = position
-        if parent_node is not None:
-            graph.add_edge(parent_node, node)
+        if node not in self._graph.nodes:
+            raise KeyError(f"No signal with source '{node}' found")
+
+    def _raise_if_data_exist(self, source_name: str):
+        """Raise an error if a data set with the passed name already exists.
+
+        Parameters
+        ----------
+        source_name :
+            Name of the data's source, e.g. a transformation or the source of the
+            measurement chain
+
+        """
+        self._raise_if_node_does_not_exist(source_name)
+        if "data" in self._graph.nodes[source_name]:
+            raise KeyError("The measurement chain already contains data for '{name}'.")
+
+    def _check_and_get_node_name(self, node_name: str) -> str:
+        """Check if a node is part of the internal graph and return its name.
+
+        If no name is provided, the name of the last added node is returned.
+
+        Parameters
+        ----------
+        node_name :
+            Name of the node that should be checked.
+
+        """
+        if node_name is None:
+            return self._prev_added_signal
+        else:
+            self._raise_if_node_does_not_exist(node_name)
+        return node_name
+
+    @staticmethod
+    def _create_signal(signal_type: str, unit: str, signal: Signal) -> Signal:
+        if signal is None:
+            if signal_type is None or unit is None:
+                raise ValueError(
+                    "You need to provide a signal type AND unit. Alternatively, an "
+                    "already existing Signal can be passed."
+                )
+            return Signal(signal_type, unit)
+        else:
+            if signal_type is not None and unit is not None:
+                warnings.warn(
+                    "Provided signal type and/or unit is ignored because an existing "
+                    "signal was passed."
+                )
+            return signal
+
+    def add_transformation(
+        self,
+        name: str,
+        error: Error,
+        output_signal_type: str = None,
+        output_signal_unit: str = None,
+        output_signal: Signal = None,
+        func: "MathematicalExpression" = None,
+        input_signal_source: str = None,
+        data=None,
+    ):
```

Comment:
So currentyl this does not create a `SignalTransformation` instance?
How about the orther way around? (attaching `SignalTransformation` to the edge?)

Reply:
I did something similar for the Signal. However, what bothers me is that attaching a `Signaltransformation` to an edge will result in storing duplicate information and therefore cluttering the asdf file. Maybe we can discuss this tomorrow. A possible solution might be to encapsulate the non-duplicate stuff in another class, but I am still struggling to see a concept that satisfies my requirements

Reply:
why would it duplicate information?
If the transformation already exists can we not just link it? (for example just passing through an existing `SignalTransformation` as the first argument)

Reply:
Because the signals are stored as nodes in the graph and inside the transformation itself. They are not necessarily the same instance, so no refs in asdf.

Reply:
So if we want to store the transformation class directly, the same should probably apply to the source. Maybe we should discuss this before I proceed.

Reply:
> 
> 
> Because the signals are stored as nodes in the graph and inside the transformation itself. They are not necessarily the same instance, so no refs in asdf.

Oh I see is it because the `SignalTransformations` have in/output signals as attributes?
```python
@dataclass
class SignalTransformation:
    name: str
    error: Error
    input_signal: Signal
    output_signal: Signal
    func: "MathematicalExpression" = None
    input_shape: Tuple = None
    output_shape: Tuple = None
```
would that not be covered by the graph now?

Reply:
It would, but only inside the graph. A transformation at a piece of equipment still needs these information. I already have an idea how to solve this. Would require some restructuring and I would really have your opinion on that @marscher @CagtayFabry before I start it. It basically affects the internal structure of the graph at the python side, not asfd: 

If we would leave the nodes blank (except for the name) and store the signal data at the transformations and the source, we have just a single duplication of each signal. They can still be different objects at the start, but we can introduce a synchronization method for that. Since signals are always encapsulated inside of an `Source` or `SignalTransformation` object, overwriting them will change the state of the encapsulating class. So lets say we have two equipments that are paired in a measurement chain. Their signals are identical but they are not same python object. When adding both to a measurement chain, we just need to copy the signal of one class into the other one to enforce an identical python object before we copy the transformation itself to the measurement chain. Does this make sense? What do you think?